flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java

```diff
@@ -0,0 +1,882 @@
+                            : null);
+        }
+    }
+
+    private void stopCheckpointServicesSafely(JobStatus terminalState) {
+        Exception exception = null;
+
+        try {
+            completedCheckpointStore.shutdown(terminalState, checkpointsCleaner);
+        } catch (Exception e) {
+            exception = e;
+        }
+
+        try {
+            checkpointIdCounter.shutdown(terminalState);
+        } catch (Exception e) {
+            exception = ExceptionUtils.firstOrSuppressed(e, exception);
+        }
+
+        if (exception != null) {
+            LOG.warn("Failed to stop checkpoint services.", exception);
+        }
+    }
+
+    @Override
+    public Executing.FailureResult howToHandleFailure(Throwable failure) {
+        if (ExecutionFailureHandler.isUnrecoverableError(failure)) {
+            return Executing.FailureResult.canNotRestart(
+                    new JobException("The failure is not recoverable", failure));
+        }
+
+        restartBackoffTimeStrategy.notifyFailure(failure);
+        if (restartBackoffTimeStrategy.canRestart()) {
+            return Executing.FailureResult.canRestart(
+                    Duration.ofMillis(restartBackoffTimeStrategy.getBackoffTime()));
+        } else {
+            return Executing.FailureResult.canNotRestart(
+                    new JobException(
+                            "Recovery is suppressed by " + restartBackoffTimeStrategy, failure));
+        }
+    }
+
+    @Override
+    public Executor getMainThreadExecutor() {
+        return componentMainThreadExecutor;
+    }
+
+    @Override
+    public boolean isState(State expectedState) {
+        return expectedState == this.state;
+    }
+
+    @Override
+    public void runIfState(State expectedState, Runnable action) {
+        if (isState(expectedState)) {
+            try {
+                action.run();
+            } catch (Throwable t) {
+                fatalErrorHandler.onFatalError(t);
+            }
+        } else {
+            LOG.debug(
+                    "Ignoring scheduled action because expected state {} is not the actual state {}.",
+                    expectedState,
+                    state);
+        }
+    }
+
+    @Override
+    public void runIfState(State expectedState, Runnable action, Duration delay) {
+        componentMainThreadExecutor.schedule(
+                () -> runIfState(expectedState, action), delay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    // ----------------------------------------------------------------
+
+    private void transitionToState(State newState) {
+        if (state != newState) {
+            LOG.debug(
+                    "Transition from state {} to {}.",
+                    state.getClass().getSimpleName(),
+                    newState.getClass().getSimpleName());
+
+            State oldState = state;
+            oldState.onLeave(newState.getClass());
+
+            state = newState;
+            newState.onEnter();
```

Comment:
Will we consider removing onEnter() separately?

Reply:
yes. As part of that I'd also like to enforce that we cannot transitions into the same state again, which should alleviate the concern about changed semantics.